File: ios/Sources/CallKitPlugin/CallKitPlugin.swift
import AVFoundation
import Capacitor
import Foundation
import UserNotifications

/// Capacitor bridge over CallKitManager / PushKitManager / AudioSessionManager.
/// Same method and event names as the Android plugin (see src/definitions.ts).
@objc(CallKitPlugin)
public class CallKitPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CallKitPlugin"
    public let jsName = "CallKit"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "setup", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "registerVoipToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "unregisterVoipToken", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "displayIncomingCall", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "answerCall", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "rejectCall", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startCall", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setCallActive", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setCallState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "endCall", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "endAllCalls", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "reportEndCall", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "updateDisplay", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setMuted", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setSpeaker", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setOnHold", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendDTMF", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getAudioRoutes", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setAudioRoute", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setReachable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setCanMakeMultipleCalls", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getInitialEvents", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearInitialEvents", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hasPhoneAccount", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openPhoneAccountSettings", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "canUseFullScreenIntent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openFullScreenIntentSettings", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "backToForeground", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getActiveCalls", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermissions", returnType: CAPPluginReturnPromise)
    ]

    private var manager: CallKitManager {
        return CallKitManager.shared
    }

    // MARK: - Lifecycle

    override public func load() {
        DispatchQueue.main.async {
            _ = CallKitManager.shared
            // Register as early as possible so VoIP pushes reach us (as the original plugin did).
            PushKitManager.shared.register()
            CallEventBus.shared.attach(owner: self) { [weak self] name, data in
                guard let self = self else {
                    return false
                }
                let listening = self.hasListeners(name)
                // Retain until a JS listener is added, so events from a cold start aren't lost.
                self.notifyListeners(name, data: data, retainUntilConsumed: true)
                return listening
            }
        }
    }

    deinit {
        CallEventBus.shared.detach(owner: self)
    }

    // MARK: - Setup / tokens

    @objc func setup(_ call: CAPPluginCall) {
        var settings: [String: Any] = [:]
        if let supportsVideo = call.getBool("supportsVideo") {
            settings["supportsVideo"] = supportsVideo
        }
        if let imageName = call.getString("imageName") {
            settings["imageName"] = imageName
        }
        let ios = call.getObject("ios") ?? [:]
        if let ringtone = ios["ringtoneSound"] as? String {
            settings["ringtoneSound"] = ringtone
        }
        for key in ["maximumCallGroups", "maximumCallsPerCallGroup"] {
            if let value = ios[key] as? NSNumber {
                settings[key] = value.intValue
            }
        }
        if let recents = ios["includesCallsInRecents"] as? Bool {
            settings["includesCallsInRecents"] = recents
        }
        if let handleType = ios["handleType"] as? String {
            settings["handleType"] = handleType
        }
        if let audio = ios["audioSession"] as? JSObject {
            var audioSettings: [String: Any] = [:]
            if let autoConfigure = audio["autoConfigure"] as? Bool {
                audioSettings["autoConfigure"] = autoConfigure
            }
            if let options = audio["categoryOptions"] as? NSNumber {
                audioSettings["categoryOptions"] = options.uintValue
            }
            if let mode = audio["mode"] as? String {
                audioSettings["mode"] = mode
            }
            settings["audioSession"] = audioSettings
        }
        // Shared with Android: `android.canMakeMultipleCalls`.
        if let android = call.getObject("android"), let allow = android["canMakeMultipleCalls"] as? Bool {
            settings["canMakeMultipleCalls"] = allow
        }

        DispatchQueue.main.async {
            self.manager.setup(settings)
            PushKitManager.shared.register()
            call.resolve()
        }
    }

    @objc func registerVoipToken(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            PushKitManager.shared.whenToken { token in
                call.resolve(["token": token])
            }
        }
    }

    @objc func unregisterVoipToken(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            PushKitManager.shared.unregister()
            call.resolve()
        }
    }

    // MARK: - Incoming

    @objc func displayIncomingCall(_ call: CAPPluginCall) {
        guard let callId = required(call, "callId"), let callerName = required(call, "callerName") else {
            return
        }
        let handle = call.getString("handle") ?? callerName
        let hasVideo = call.getBool("hasVideo") ?? false
        let payload = call.getObject("payload")?.mapValues { $0 as Any }
        DispatchQueue.main.async {
            self.manager.reportIncomingCall(callId: callId, callerName: callerName, handle: handle, hasVideo: hasVideo, payload: payload) {
                self.finish(call, $0)
            }
        }
    }

    @objc func answerCall(_ call: CAPPluginCall) {
        guard let callId = required(call, "callId") else {
            return
        }
        DispatchQueue.main.async {
            self.manager.answer(callId) { self.finish(call, $0) }
        }
    }

    @objc func rejectCall(_ call: CAPPluginCall) {
        guard let callId = required(call, "callId") else {
            return
        }
        DispatchQueue.main.async {
            self.manager.reject(callId) { self.finish(call, $0) }
        }
    }

    // MARK: - Outgoing

    @objc func startCall(_ call: CAPPluginCall) {
        guard let callId = required(call, "callId"), let calleeName = required(call, "calleeName") else {
            return
        }
        let handle = call.getString("handle") ?? calleeName
        let hasVideo = call.getBool("hasVideo") ?? false
        DispatchQueue.main.async {
            self.manager.startCall(callId: callId, calleeName: calleeName, handle: handle, hasVideo: hasVideo) {
                self.finish(call, $0)
            }
        }
    }

    @objc func setCallActive(_ call: CAPPluginCall) {
        guard let callId = required(call, "callId") else {
            return
        }
        run(call) { try self.manager.setCallActive(callId) }
    }

    @objc func setCallState(_ call: CAPPluginCall) {
        guard let callId = required(call, "callId"), let state = required(call, "state") else {
            return
        }
        DispatchQueue.main.async {
            self.manager.setCallState(callId, state: state) { self.finish(call, $0) }
        }
    }

    @objc func endCall(_ call: CAPPluginCall) {
        guard let callId = required(call, "callId") else {
            return
        }
        DispatchQueue.main.async {
            self.manager.endCall(callId) { self.finish(call, $0) }
        }
    }

    @objc func endAllCalls(_ call: CAPPluginCall) {
        run(call) { self.manager.endAllCalls() }
    }

    @objc func reportEndCall(_ call: CAPPluginCall) {
        guard let callId = required(call, "callId") else {
            return
        }
        let reason = call.getInt("reason") ?? 2
        run(call) { try self.manager.reportEndCall(callId, reason: reason) }
    }

    @objc func updateDisplay(_ call: CAPPluginCall) {
        guard let callId = required(call, "callId"), let callerName = required(call, "callerName") else {
            return
        }
        let handle = call.getString("handle") ?? callerName
        run(call) { try self.manager.updateDisplay(callId, callerName: callerName, handle: handle) }
    }

    // MARK: - In-call controls

    @objc func setMuted(_ call: CAPPluginCall) {
        guard let callId = required(call, "callId") else {
            return
        }
        let muted = call.getBool("muted") ?? false
        DispatchQueue.main.async {
            self.manager.setMuted(callId, muted: muted) { self.finish(call, $0) }
        }
    }

    @objc func setSpeaker(_ call: CAPPluginCall) {
        let on = call.getBool("on") ?? false
        // The route-change notification emits speakerChanged / audioRouteChanged.
        run(call) { try AudioSessionManager.shared.setSpeaker(on) }
    }

    @objc func setOnHold(_ call: CAPPluginCall) {
        guard let callId = required(call, "callId") else {
            return
        }
        let hold = call.getBool("hold") ?? false
        DispatchQueue.main.async {
            self.manager.setOnHold(callId, hold: hold) { self.finish(call, $0) }
        }
    }

    @objc func sendDTMF(_ call: CAPPluginCall) {
        guard let callId = required(call, "callId"), let digits = required(call, "digits") else {
            return
        }
        DispatchQueue.main.async {
            self.manager.sendDTMF(callId, digits: digits) { self.finish(call, $0) }
        }
    }

    @objc func getAudioRoutes(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            do {
                call.resolve(["routes": try AudioSessionManager.shared.audioRoutes()])
            } catch {
                call.reject("GetAudioRoutes Error: \(error.localizedDescription)")
            }
        }
    }

    @objc func setAudioRoute(_ call: CAPPluginCall) {
        guard let route = required(call, "route") else {
            return
        }
        run(call) { try AudioSessionManager.shared.setRoute(route) }
    }

    // MARK: - Platform helpers

    @objc func setAvailable(_ call: CAPPluginCall) {
        call.resolve() // Android only.
    }

    @objc func setReachable(_ call: CAPPluginCall) {
        call.resolve() // Android only.
    }

    @objc func setCanMakeMultipleCalls(_ call: CAPPluginCall) {
        let allow = call.getBool("allow") ?? true
        run(call) { self.manager.setCanMakeMultipleCalls(allow) }
    }

    @objc func getInitialEvents(_ call: CAPPluginCall) {
        call.resolve(["events": CallEventBus.shared.initialEvents().map { $0.toDictionary() }])
    }

    @objc func clearInitialEvents(_ call: CAPPluginCall) {
        CallEventBus.shared.clearInitialEvents()
        call.resolve()
    }

    @objc func hasPhoneAccount(_ call: CAPPluginCall) {
        call.resolve(["value": true]) // No calling account on iOS.
    }

    @objc func openPhoneAccountSettings(_ call: CAPPluginCall) {
        call.unimplemented("Not available on iOS.")
    }

    @objc func canUseFullScreenIntent(_ call: CAPPluginCall) {
        call.resolve(["value": true]) // CallKit always shows the full-screen call UI.
    }

    @objc func openFullScreenIntentSettings(_ call: CAPPluginCall) {
        call.unimplemented("Not available on iOS.")
    }

    @objc func backToForeground(_ call: CAPPluginCall) {
        call.unimplemented("iOS does not let apps bring themselves to the foreground.")
    }

    @objc func getActiveCalls(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            call.resolve(["calls": self.manager.activeCalls()])
        }
    }

    // MARK: - Permissions

    @objc override public func checkPermissions(_ call: CAPPluginCall) {
        permissionStatus { call.resolve($0) }
    }

    @objc override public func requestPermissions(_ call: CAPPluginCall) {
        let wanted = call.getArray("permissions", String.self) ?? ["microphone", "notifications"]
        let requestNotifications = {
            guard wanted.contains("notifications") else {
                self.permissionStatus { call.resolve($0) }
                return
            }
            UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound, .badge]) { _, _ in
                self.permissionStatus { call.resolve($0) }
            }
        }
        if wanted.contains("microphone") {
            AVAudioSession.sharedInstance().requestRecordPermission { _ in
                requestNotifications()
            }
        } else {
            requestNotifications()
        }
    }

    private func permissionStatus(_ done: @escaping ([String: Any]) -> Void) {
        let microphone: String
        switch AVAudioSession.sharedInstance().recordPermission {
        case .granted:
            microphone = "granted"
        case .denied:
            microphone = "denied"
        default:
            microphone = "prompt"
        }
        UNUserNotificationCenter.current().getNotificationSettings { settings in
            let notifications: String
            switch settings.authorizationStatus {
            case .authorized, .provisional, .ephemeral:
                notifications = "granted"
            case .denied:
                notifications = "denied"
            default:
                notifications = "prompt"
            }
            done(["microphone": microphone, "notifications": notifications, "phone": "granted"])
        }
    }

    // MARK: - Helpers

    private func required(_ call: CAPPluginCall, _ key: String) -> String? {
        guard let value = call.getString(key), !value.isEmpty else {
            call.reject("Missing required option: \(key)")
            return nil
        }
        return value
    }

    private func run(_ call: CAPPluginCall, _ action: @escaping () throws -> Void) {
        DispatchQueue.main.async {
            do {
                try action()
                call.resolve()
            } catch {
                call.reject(error.localizedDescription)
            }
        }
    }

    private func finish(_ call: CAPPluginCall, _ error: Error?) {
        if let error = error {
            call.reject(error.localizedDescription)
        } else {
            call.resolve()
        }
    }
}

File: android/src/main/java/io/kreador/callkit/CallKitPlugin.java
package io.kreador.callkit;

import static io.kreador.callkit.Constants.EXTRA_CALL_UUID;
import static io.kreador.callkit.Constants.EXTRA_LAUNCH_ACTION;
import static io.kreador.callkit.Constants.LAUNCH_ACTION_ANSWER;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.os.Build;
import android.util.Log;
import android.view.WindowManager;
import androidx.core.app.NotificationManagerCompat;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;
import com.google.firebase.messaging.FirebaseMessaging;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import org.json.JSONException;
import org.json.JSONObject;

/**
 * Capacitor bridge over CallKeepModule (the callkeep port). Replaces callkeep's
 * @ReactMethod surface with @PluginMethod and its event emitter with notifyListeners.
 */
@CapacitorPlugin(
    name = "CallKit",
    permissions = {
        @Permission(alias = CallKitPlugin.MICROPHONE, strings = { Manifest.permission.RECORD_AUDIO }),
        @Permission(alias = CallKitPlugin.NOTIFICATIONS, strings = { Manifest.permission.POST_NOTIFICATIONS }),
        @Permission(
            alias = CallKitPlugin.PHONE,
            strings = { Manifest.permission.READ_PHONE_STATE, Manifest.permission.READ_PHONE_NUMBERS, Manifest.permission.CALL_PHONE }
        )
    }
)
public class CallKitPlugin extends Plugin implements CallEventBus.Listener {

    private static final String TAG = "IonicCallkit";
    static final String MICROPHONE = "microphone";
    static final String NOTIFICATIONS = "notifications";
    static final String PHONE = "phone";

    private CallKeepModule module;

    private interface Action {
        void run() throws Exception;
    }

    // ─── Lifecycle ──────────────────────────────────────────────────────────

    @Override
    public void load() {
        module = CallKeepModule.getInstance(getContext());
        module.setCurrentActivity(getActivity());
        CallEventBus.attach(this);
        if (getActivity() != null) {
            handleLaunchIntent(getActivity().getIntent());
        }
    }

    @Override
    protected void handleOnNewIntent(Intent intent) {
        super.handleOnNewIntent(intent);
        handleLaunchIntent(intent);
    }

    @Override
    protected void handleOnResume() {
        super.handleOnResume();
        module.setCurrentActivity(getActivity());
    }

    @Override
    protected void handleOnDestroy() {
        super.handleOnDestroy();
        CallEventBus.detach(this);
        CallConnectionService.setUnreachable();
        Activity activity = getActivity();
        if (activity != null && activity.isFinishing() && !activity.isChangingConfigurations()) {
            module.setCurrentActivity(null);
            module.onHostDestroy();
        }
    }

    /**
     * A `callAnswered` listener means the app can take over an answered call, so it counts
     * as setReachable() for the answer watchdog — apps that forget setReachable() don't
     * get their calls ended.
     */
    @Override
    @PluginMethod(returnType = PluginMethod.RETURN_NONE)
    public void addListener(PluginCall call) {
        super.addListener(call);
        if ("callAnswered".equals(call.getString("eventName"))) {
            CallConnectionService.setReachable();
        }
    }

    @Override
    public boolean onCallEvent(String eventName, JSObject data) {
        boolean jsListening = hasListeners(eventName);
        if ("callEnded".equals(eventName) || "callRejected".equals(eventName)) {
            if (CallConnectionService.currentConnections.isEmpty()) {
                setShowOverLockScreen(false);
            }
        }
        // Retain until a JS listener is added, so events from a cold start aren't lost.
        notifyListeners(eventName, data, true);
        return jsListening;
    }

    /** Notification taps (full-screen intent / Answer) launch the app with these extras. */
    private void handleLaunchIntent(Intent intent) {
        if (intent == null || !intent.hasExtra(EXTRA_LAUNCH_ACTION)) {
            return;
        }
        String action = intent.getStringExtra(EXTRA_LAUNCH_ACTION);
        String callId = intent.getStringExtra(EXTRA_CALL_UUID);
        // Consume, so an activity recreation doesn't answer twice.
        intent.removeExtra(EXTRA_LAUNCH_ACTION);

        Log.d(TAG, "[CallKitPlugin] launch action: " + action + ", uuid: " + callId);
        setShowOverLockScreen(true);

        if (LAUNCH_ACTION_ANSWER.equals(action) && callId != null) {
            try {
                module.answerIncomingCall(callId);
            } catch (CallKeepModule.CallKeepException e) {
                Log.w(TAG, "[CallKitPlugin] answer from notification: " + e.getMessage());
                IncomingCallNotification.cancel(getContext(), callId);
            }
        }
    }

    private void setShowOverLockScreen(boolean show) {
        Activity activity = getActivity();
        if (activity == null) {
            return;
        }
        activity.runOnUiThread(() -> {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
                activity.setShowWhenLocked(show);
                activity.setTurnScreenOn(show);
            } else {
                // Android 8.0 has no setShowWhenLocked / setTurnScreenOn.
                int flags = WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED | WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON;
                if (show) {
                    activity.getWindow().addFlags(flags);
                } else {
                    activity.getWindow().clearFlags(flags);
                }
            }
        });
    }

    // ─── Setup / tokens ─────────────────────────────────────────────────────

    @PluginMethod
    public void setup(PluginCall call) {
        run(call, () -> {
            JSObject data = call.getData();
            JSONObject settings = new JSONObject();
            copy(data, settings, "appName", "imageName", "supportsVideo");
            JSONObject android = data.optJSONObject("android");
            if (android != null) {
                Iterator<String> keys = android.keys();
                while (keys.hasNext()) {
                    String key = keys.next();
                    settings.put(key, android.get(key));
                }
            }
            module.setup(settings);
        });
    }

    @PluginMethod
    public void registerVoipToken(PluginCall call) {
        FirebaseMessaging messaging;
        try {
            messaging = FirebaseMessaging.getInstance();
        } catch (IllegalStateException e) {
            call.reject("Firebase is not initialised. Add google-services.json to your Android app.", e);
            return;
        }
        messaging
            .getToken()
            .addOnCompleteListener(task -> {
                if (!task.isSuccessful() || task.getResult() == null) {
                    Exception e = task.getException();
                    call.reject("Could not get FCM token: " + (e != null ? e.getMessage() : "unknown error"), e);
                    return;
                }
                String token = task.getResult();
                JSObject ret = new JSObject();
                ret.put("token", token);
                call.resolve(ret);
                module.onNewToken(token);
            });
    }

    @PluginMethod
    public void unregisterVoipToken(PluginCall call) {
        try {
            FirebaseMessaging.getInstance()
                .deleteToken()
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        call.resolve();
                    } else {
                        call.reject("Could not delete FCM token", task.getException());
                    }
                });
        } catch (IllegalStateException e) {
            call.reject("Firebase is not initialised.", e);
        }
    }

    // ─── Incoming ───────────────────────────────────────────────────────────

    @PluginMethod
    public void displayIncomingCall(PluginCall call) {
        String callId = required(call, "callId");
        String callerName = required(call, "callerName");
        if (callId == null || callerName == null) {
            return;
        }
        String handle = call.getString("handle", callerName);
        boolean hasVideo = Boolean.TRUE.equals(call.getBoolean("hasVideo", false));
        JSObject payload = call.getObject("payload", null);
        run(call, () -> module.reportNewIncomingCall(callId, handle, callerName, hasVideo, payload));
    }

    @PluginMethod
    public void answerCall(PluginCall call) {
        String callId = required(call, "callId");
        if (callId == null) return;
        run(call, () -> module.answerIncomingCall(callId));
    }

    @PluginMethod
    public void rejectCall(PluginCall call) {
        String callId = required(call, "callId");
        if (callId == null) return;
        run(call, () -> module.rejectCall(callId));
    }

    // ─── Outgoing ───────────────────────────────────────────────────────────

    @PluginMethod
    public void startCall(PluginCall call) {
        String callId = required(call, "callId");
        String calleeName = required(call, "calleeName");
        if (callId == null || calleeName == null) {
            return;
        }
        String handle = call.getString("handle", calleeName);
        boolean hasVideo = Boolean.TRUE.equals(call.getBoolean("hasVideo", false));
        run(call, () -> module.startCall(callId, handle, calleeName, hasVideo));
    }

    @PluginMethod
    public void setCallActive(PluginCall call) {
        String callId = required(call, "callId");
        if (callId == null) return;
        run(call, () -> module.setCurrentCallActive(callId));
    }

    @PluginMethod
    public void endCall(PluginCall call) {
        String callId = required(call, "callId");
        if (callId == null) return;
        run(call, () -> module.endCall(callId));
    }

    @PluginMethod
    public void endAllCalls(PluginCall call) {
        run(call, () -> module.endAllCalls());
    }

    @PluginMethod
    public void reportEndCall(PluginCall call) {
        String callId = required(call, "callId");
        if (callId == null) return;
        int reason = call.getInt("reason", 2);
        run(call, () -> module.reportEndCallWithUUID(callId, reason));
    }

    @PluginMethod
    public void updateDisplay(PluginCall call) {
        String callId = required(call, "callId");
        String callerName = required(call, "callerName");
        if (callId == null || callerName == null) {
            return;
        }
        String handle = call.getString("handle", callerName);
        run(call, () -> module.updateDisplay(callId, callerName, handle));
    }

    // ─── In-call controls ───────────────────────────────────────────────────

    @PluginMethod
    public void setMuted(PluginCall call) {
        String callId = required(call, "callId");
        if (callId == null) return;
        boolean muted = Boolean.TRUE.equals(call.getBoolean("muted", false));
        run(call, () -> module.setMutedCall(callId, muted));
    }

    @PluginMethod
    public void setSpeaker(PluginCall call) {
        boolean on = Boolean.TRUE.equals(call.getBoolean("on", false));
        String callId = call.getString("callId");
        run(call, () -> module.toggleAudioRouteSpeaker(callId, on));
    }

    @PluginMethod
    public void setOnHold(PluginCall call) {
        String callId = required(call, "callId");
        if (callId == null) return;
        boolean hold = Boolean.TRUE.equals(call.getBoolean("hold", false));
        run(call, () -> module.setOnHold(callId, hold));
    }

    @PluginMethod
    public void sendDTMF(PluginCall call) {
        String callId = required(call, "callId");
        String digits = required(call, "digits");
        if (callId == null || digits == null) {
            return;
        }
        run(call, () -> module.sendDTMF(callId, digits));
    }

    @PluginMethod
    public void getAudioRoutes(PluginCall call) {
        try {
            JSObject ret = new JSObject();
            ret.put("routes", module.getAudioRoutes());
            call.resolve(ret);
        } catch (Exception e) {
            call.reject("GetAudioRoutes Error: " + e.getMessage(), e);
        }
    }

    @PluginMethod
    public void setAudioRoute(PluginCall call) {
        String callId = required(call, "callId");
        String route = required(call, "route");
        if (callId == null || route == null) {
            return;
        }
        run(call, () -> module.setAudioRoute(callId, route));
    }

    // ─── Android specifics ──────────────────────────────────────────────────

    @PluginMethod
    public void setAvailable(PluginCall call) {
        boolean available = Boolean.TRUE.equals(call.getBoolean("available", true));
        run(call, () -> module.setAvailable(available));
    }

    @PluginMethod
    public void setReachable(PluginCall call) {
        run(call, () -> module.setReachable());
    }

    @PluginMethod
    public void setCallState(PluginCall call) {
        String callId = required(call, "callId");
        String state = required(call, "state");
        if (callId == null || state == null) {
            return;
        }
        run(call, () -> module.setCallState(callId, state));
    }

    @PluginMethod
    public void setCanMakeMultipleCalls(PluginCall call) {
        boolean allow = Boolean.TRUE.equals(call.getBoolean("allow", true));
        run(call, () -> module.setCanMakeMultipleCalls(allow));
    }

    @PluginMethod
    public void getInitialEvents(PluginCall call) {
        JSArray events = new JSArray();
        for (CallEventBus.Event event : CallEventBus.getInitialEvents()) {
            events.put(event.toJSObject());
        }
        JSObject ret = new JSObject();
        ret.put("events", events);
        call.resolve(ret);
    }

    @PluginMethod
    public void clearInitialEvents(PluginCall call) {
        CallEventBus.clearInitialEvents();
        call.resolve();
    }

    @PluginMethod
    public void hasPhoneAccount(PluginCall call) {
        resolveValue(call, module.hasPhoneAccount());
    }

    @PluginMethod
    public void openPhoneAccountSettings(PluginCall call) {
        run(call, () -> module.openPhoneAccounts());
    }

    @PluginMethod
    public void canUseFullScreenIntent(PluginCall call) {
        resolveValue(call, module.canUseFullScreenIntent());
    }

    @PluginMethod
    public void openFullScreenIntentSettings(PluginCall call) {
        run(call, () -> module.openFullScreenIntentSettings());
    }

    @PluginMethod
    public void backToForeground(PluginCall call) {
        run(call, () -> module.backToForeground());
    }

    @PluginMethod
    public void getActiveCalls(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("calls", module.getActiveCalls());
        call.resolve(ret);
    }

    // ─── Permissions ────────────────────────────────────────────────────────

    @Override
    @PluginMethod
    public void checkPermissions(PluginCall call) {
        call.resolve(permissionStatus());
    }

    @Override
    @PluginMethod
    public void requestPermissions(PluginCall call) {
        List<String> aliases = new ArrayList<>();
        JSArray requested = call.getArray("permissions");
        if (requested != null && requested.length() > 0) {
            try {
                aliases.addAll(requested.<String>toList());
            } catch (JSONException e) {
                call.reject("permissions must be an array of strings");
                return;
            }
        } else {
            aliases.add(MICROPHONE);
            aliases.add(NOTIFICATIONS);
            if (!module.isSelfManaged()) {
                aliases.add(PHONE);
            }
        }
        // POST_NOTIFICATIONS only exists as a runtime permission on Android 13+.
        if (Build.VERSION.SDK_INT < 33) {
            aliases.remove(NOTIFICATIONS);
        }
        if (aliases.isEmpty()) {
            call.resolve(permissionStatus());
            return;
        }
        requestPermissionForAliases(aliases.toArray(new String[0]), call, "permissionsCallback");
    }

    @PermissionCallback
    private void permissionsCallback(PluginCall call) {
        call.resolve(permissionStatus());
    }

    private JSObject permissionStatus() {
        JSObject result = new JSObject();
        result.put(MICROPHONE, getPermissionState(MICROPHONE).toString());
        if (Build.VERSION.SDK_INT < 33) {
            boolean enabled = NotificationManagerCompat.from(getContext()).areNotificationsEnabled();
            result.put(NOTIFICATIONS, (enabled ? PermissionState.GRANTED : PermissionState.DENIED).toString());
        } else {
            result.put(NOTIFICATIONS, getPermissionState(NOTIFICATIONS).toString());
        }
        result.put(PHONE, getPermissionState(PHONE).toString());
        return result;
    }

    // ─── Helpers ────────────────────────────────────────────────────────────

    private void run(PluginCall call, Action action) {
        try {
            action.run();
            call.resolve();
        } catch (CallKeepModule.CallKeepException e) {
            call.reject(e.getMessage());
        } catch (Exception e) {
            Log.e(TAG, "[CallKitPlugin] " + call.getMethodName() + " failed", e);
            call.reject(e.getMessage() != null ? e.getMessage() : e.toString(), e);
        }
    }

    private String required(PluginCall call, String key) {
        String value = call.getString(key);
        if (value == null || value.isEmpty()) {
            call.reject("Missing required option: " + key);
            return null;
        }
        return value;
    }

    private void resolveValue(PluginCall call, boolean value) {
        JSObject ret = new JSObject();
        ret.put("value", value);
        call.resolve(ret);
    }

    private static void copy(JSONObject from, JSONObject to, String... keys) throws JSONException {
        for (String key : keys) {
            if (from.has(key)) {
                to.put(key, from.get(key));
            }
        }
    }
}
